crates/context/config/src/client/protocol/icp.rs

```diff
@@ -0,0 +1,262 @@
+            return Err(IcpError::UnknownNetwork(request.network_id.into_owned()));
+        };
+
+        let canister_id = Principal::from_text(&request.contract_id)
+            .map_err(|_| IcpError::InvalidCanisterId(request.contract_id.into_owned()))?;
+
+        match request.operation {
+            Operation::Read { method } => {
+                network
+                    .query(&canister_id, method.into_owned(), payload)
+                    .await
+            }
+            Operation::Write { method } => {
+                network
+                    .mutate(&canister_id, method.into_owned(), payload)
+                    .await
+            }
+        }
+    }
+}
+
+impl Network {
+    async fn query(
+        &self,
+        canister_id: &Principal,
+        method: String,
+        args: Vec<u8>,
+    ) -> Result<Vec<u8>, IcpError> {
+        self.client
+            .fetch_root_key()
+            .await
+            .map_err(|_| IcpError::Custom {
+                operation: ErrorOperation::Query,
+                reason: "Failed to fetch root key".to_string(),
+            })?;
+
+        let response = self
+            .client
+            .query(canister_id, method)
+            .with_arg(args)
+            .call()
+            .await;
+
+        response.map_or(
+            Err(IcpError::Custom {
+                operation: ErrorOperation::Query,
+                reason: "Error while quering".to_string(),
+            }),
+            |response| Ok(response),
+        )
+    }
+
+    async fn mutate(
+        &self,
+        canister_id: &Principal,
+        method: String,
+        args: Vec<u8>,
+    ) -> Result<Vec<u8>, IcpError> {
+        self.client
+            .fetch_root_key()
+            .await
+            .map_err(|_| IcpError::Custom {
+                operation: ErrorOperation::Query,
+                reason: "Failed to fetch root key".to_string(),
+            })?;
+
+        let response = self
+            .client
+            .update(canister_id, method)
+            .with_arg(args)
+            .call()
+            .await;
+
+        match response {
+            Ok(CallResponse::Response((data, _))) => {
+                // Now try to decode the actual data
+                match candid::decode_one::<Result<(), String>>(&data) {
+                    Ok(decoded) => {
+                        match decoded {
+                            Ok(()) => Ok(vec![]),                     // Return empty vec for success
+                            Err(err_msg) => Ok(err_msg.into_bytes()), // Return error message as bytes
+                        }
+                    }
+                    Err(e) => {
+                        println!("Failed to decode: {}", e);
+                        Ok(e.to_string().into_bytes()) // Return decode error as bytes
+                    }
+                }
+            }
+            Ok(CallResponse::Poll(_)) => Ok("Unexpected polling response".as_bytes().to_vec()),
```

Comment:
this concerns me, should we be checking the status in a loop?

Reply:
what does it mean to get this response though? can you link me to the doc?

Reply:
https://internetcomputer.org/docs/current/developer-docs/developer-tools/off-chain/agents/rust-agent#example 
https://internetcomputer.org/docs/current/references/ingress-messages#recommendations-and-best-practices-for-secure-update-call-handling
https://internetcomputer.org/docs/current/references/ic-interface-spec#http-async-call-overview

"Related to the asynchronous nature of update calls on the Internet Computer. This response indicates that the call is still being processed and requires further polling to get the final result."

Okay so maybe update this to use ".call_and_wait()" and then remove this option  
cc @alenmestrov ?

Reply:
Update to remove polling